test/tools/catch-analyzer/src/main/java/org/elasticsearch/test/CatchAnalyzer.java

```diff
@@ -0,0 +1,154 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.Opcodes;
+
+/**
+ * Analyzes methods for broken catch blocks. These are catch blocks that somehow
+ * drop the original exception (via any possible codepath). It is ok to wrap the exception
+ * with another one, e.g.:
+ * <ul>
+ *   <li>new OtherException(..., original, ...)
+ *   <li>otherException.initCause(original)
+ *   <li>otherException.addSuppressed(original)
+ * </ul>
+ */
+public class CatchAnalyzer {
+ 
+    /** Exception thrown when some exceptions are swallowed */
+    @SuppressWarnings("serial")
+    public static class SwallowedException extends RuntimeException {
+        public SwallowedException(String message) {
+            super(message);
+        }
+    }
+    
+    static final int ASM_API_VERSION = Opcodes.ASM6;
+    
+    /** Main method, takes directories as parameter. These must also be on classpath!!!! */
+    public static void main(String args[]) throws Exception {
+        long violationCount = 0;
+        long scannedCount = 0;
+        long startTime = System.currentTimeMillis();
+        List<Path> files = new ArrayList<>();
+        if (args.length < 3) {
+            throw new IllegalArgumentException("CatchAnalyzer -classpath <path> <dirs>");
+        }
+        if (args[0].equals("-classpath") == false) {
+            throw new IllegalArgumentException("CatchAnalyzer -classpath <path> <dirs>");
+        }
+        
+        ClassLoader loader = createLoader(args[1]);
+
+        // step 1: collect files
+        for (int i = 2; i < args.length; i++) {
+            String arg = args[i];
+            Path dir = Paths.get(arg);
+            Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    if (file.toString().endsWith(".class")) {
+                        files.add(file);
+                    }
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        }
+        // step 2: sort
+        files.sort((x,y) -> x.toAbsolutePath().toString().compareTo(y.toAbsolutePath().toString()));
+        // step 3: process
+        for (Path file : files) {
+            byte bytes[] = Files.readAllBytes(file);
+            ClassReader reader = new ClassReader(bytes);
+            // entirely synthetic class, e.g. enum switch table, which always masks NoSuchFieldError!!!!!
+            if ((reader.getAccess() & Opcodes.ACC_SYNTHETIC) != 0) {
+                continue;
+            }
+            ClassAnalyzer analyzer = new ClassAnalyzer(reader.getClassName(), loader);
+            reader.accept(analyzer, 0);
+            scannedCount++;
+            
+            violationCount += analyzer.logViolations(System.out);
+        }
+        long endTime = System.currentTimeMillis();
```

Comment:
It is probably better to use nanoTime here so you don't get clock skew giving you weird numbers. Not like it matters a whole lot here though.